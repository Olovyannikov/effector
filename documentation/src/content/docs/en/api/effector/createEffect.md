---
title: createEffect
description: Method for creating an effect
redirectFrom:
  - /api/effector/createEffect
  - /docs/api/effector/createEffect
---

# createEffect (#create-effect)

```ts
import { createEffect } from "effector";

const effectFx = createEffect();
```

Method for creating [effects](/en/api/effector/Effect). Returns a new [effect](/en/api/effector/Effect).

## How to Create Effects (#how-to-create-effects)

The `createEffect` method supports several ways to create effects:

1. With a handler - this is the simplest way.
2. With configuration.
3. Without a handler, which can be set later using the [`.use(handler)`](/en/api/effector/Effect#use-method) method.

### With Handler (#create-with-handler)

- **Type**

```ts
createEffect<Params, Done, Fail = Error>(
  handler: (params: Params) => Done | Promise<Done>,
): Effect<Params, Done, Fail>
```

- **Example**

```ts
import { createEffect } from "effector";

const fetchUserReposFx = createEffect(async ({ name }) => {
  const url = `https://api.github.com/users/${name}/repos`;
  const req = await fetch(url);
  return req.json();
});

fetchUserReposFx.done.watch(({ params, result }) => {
  console.log(result);
});

await fetchUserReposFx({ name: "zerobias" });
```

### With Configuration (#create-with-config)

The `name` field is used to improve error messages and debugging.

- **Type**

```ts
export function createEffect<Params, Done, Fail = Error>(config: {
  name?: string;
  handler?: (params: Params) => Promise<Done> | Done;
}): Effect<Params, Done, Fail>;
```

- **Example**

```ts
import { createEffect } from "effector";

const fetchUserReposFx = createEffect({
  name: "fetch user repositories",
  async handler({ name }) {
    const url = `https://api.github.com/users/${name}/repos`;
    const req = await fetch(url);
    return req.json();
  },
});

await fetchUserReposFx({ name: "zerobias" });
```

### Without Handler (#create-without-handler)

Most commonly used for testing. [More detailed information](/en/api/effector/Effect#use-method).

:::warning{title="use is an anti-pattern"}
Try to avoid using `.use()`, as it's an anti-pattern and degrades type inference.
:::

- **Example**

```ts
import { createEffect } from "effector";

const fetchUserReposFx = createEffect();

fetchUserReposFx.use(async ({ name }) => {
  const url = `https://api.github.com/users/${name}/repos`;
  const req = await fetch(url);
  return req.json();
});

await fetchUserReposFx({ name: "zerobias" });
```

## Examples (#examples)

- **Updating state on effect completion**:

```ts
import { createStore, createEffect } from "effector";

interface Repo {
  // ...
}

const $repos = createStore<Repo[]>([]);

const fetchUserReposFx = createEffect(async (name: string) => {
  const url = `https://api.github.com/users/${name}/repos`;
  const req = await fetch(url);
  return req.json();
});

$repos.on(fetchUserReposFx.doneData, (_, repos) => repos);

$repos.watch((repos) => {
  console.log(`${repos.length} repos`);
});
// => 0 repos

await fetchUserReposFx("zerobias");
// => 26 repos
```

[Run example](https://share.effector.dev/uAJFC1XM)

- **Watching effect state**:

```js
import { createEffect } from "effector";

const fetchUserReposFx = createEffect(async ({ name }) => {
  const url = `https://api.github.com/users/${name}/repos`;
  const req = await fetch(url);
  return req.json();
});

fetchUserReposFx.pending.watch((pending) => {
  console.log(`effect is pending?: ${pending ? "yes" : "no"}`);
});

fetchUserReposFx.done.watch(({ params, result }) => {
  console.log(params); // {name: 'zerobias'}
  console.log(result); // resolved value, result
});

fetchUserReposFx.fail.watch(({ params, error }) => {
  console.error(params); // {name: 'zerobias'}
  console.error(error); // rejected value, error
});

fetchUserReposFx.finally.watch(({ params, status, result, error }) => {
  console.log(params); // {name: 'zerobias'}
  console.log(`handler status: ${status}`);

  if (error) {
    console.log("handler rejected", error);
  } else {
    console.log("handler resolved", result);
  }
});

await fetchUserReposFx({ name: "zerobias" });
```

[Run example](https://share.effector.dev/LeurvtYA)

## Related API and Articles (#related-api-and-docs-to-create-effect)

- **API**
  - [`Effect API`](/en/api/effector/Effect) - Description of effects, their methods and properties
  - [`sample`](/en/api/effector/sample) - Key operator for building connections between units
  - [`attach`](/en/api/effector/attach) - Creates new effects based on other effects
- **Articles**
  - [Working with effects](/en/essentials/work-with-async)
  - [How to type effects and other units](/en/essentials/typescript)
  - [Guide to testing effects and other units](/en/guides/testing)
