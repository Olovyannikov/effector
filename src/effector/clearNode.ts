import {Node, NodeUnit} from './index.h'
import {getGraph, getOwners, getLinks, getSubscribers} from './getter'
import {is} from './is'
import {removeItem} from './collection'
import {CROSSLINK} from './tag'

const removeFromNode = (currentNode: Node, targetNode: Node) => {
  removeItem(currentNode.next, targetNode)
  removeItem(getOwners(currentNode), targetNode)
  removeItem(getLinks(currentNode), targetNode)
}
const clearNodeNormalized = (
  targetNode: Node,
  deep: boolean,
  isDomainUnit: boolean,
  regionNode: Node | null,
  extractOnly: boolean,
) => {
  targetNode.next.length = 0
  targetNode.seq.length = 0
  //@ts-expect-error
  targetNode.scope = null
  let currentNode
  let list = getLinks(targetNode)
  const isRegionNode = targetNode.meta.isRegion
  const nextRegionNode = isRegionNode ? targetNode : regionNode
  if (list.length > 0) {
    const canGoDeep = !isRegionNode && !extractOnly
    const domainSampleEdgeCase =
      canGoDeep && isDomainUnit && targetNode.meta.op !== 'sample'
    while ((currentNode = list.pop())) {
      removeFromNode(currentNode, targetNode)
      if (isRegionNode) {
        clearNodeNormalized(currentNode, false, false, targetNode, true)
      }
      if (
        deep ||
        domainSampleEdgeCase ||
        (canGoDeep && currentNode.family.type === CROSSLINK)
      ) {
        clearNodeNormalized(
          currentNode,
          deep,
          isDomainUnit && currentNode.meta.op !== 'on',
          nextRegionNode,
          extractOnly,
        )
      }
    }
  }
  list = getOwners(targetNode)
  while ((currentNode = list.pop())) {
    removeFromNode(currentNode, targetNode)
    if (isDomainUnit && currentNode.family.type === CROSSLINK) {
      clearNodeNormalized(
        currentNode,
        deep,
        currentNode.meta.op !== 'on',
        nextRegionNode,
        extractOnly,
      )
    }
  }
}
const clearMap = (map: Map<any, any> | Set<any>) => map.clear()
export const clearNode = (
  graphite: NodeUnit,
  {
    deep,
  }: {
    deep?: boolean
  } = {},
) => {
  let isDomainUnit = false
  //@ts-expect-error
  if (graphite.ownerSet) graphite.ownerSet.delete(graphite)
  if (is.store(graphite)) {
    clearMap(getSubscribers(graphite))
  } else if (is.domain(graphite)) {
    isDomainUnit = true
    const history = graphite.history
    clearMap(history.events)
    clearMap(history.effects)
    clearMap(history.stores)
    clearMap(history.domains)
  }
  clearNodeNormalized(getGraph(graphite), !!deep, isDomainUnit, null, false)
}
