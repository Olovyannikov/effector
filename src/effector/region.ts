import type {Template} from '../forest/index.h'
import type {NodeUnit, Node, ID} from './index.h'
import {getParent, getGraph} from './getter'
import {createNode} from './createNode'
import {is} from './is'

type DeclarationSourceReporter = (
  node: Node | 'region',
  regionStack: RegionStack | null,
) => void

let reporter: DeclarationSourceReporter

export const setGraphInspector = (fn: DeclarationSourceReporter) => {
  reporter = fn
}

type RegionStack = {
  id: ID
  parent: RegionStack | null
  value: any
  template: Template | null
  sidRoot?: string
  meta:
    | Record<string, unknown>
    | {
        type: 'factory'
        sid?: string
        name?: string
        loc: unknown
        method?: string
      }
}

export let regionStack: RegionStack | null = null

export const reportDeclaration = (node: Node | 'region') => {
  if (reporter) {
    reporter(node, regionStack)
  }
}

export const readTemplate = (): Template | null =>
  regionStack && regionStack.template
export const readSidRoot = (sid?: string | null) => {
  if (sid && regionStack && regionStack.sidRoot)
    sid = `${regionStack.sidRoot}|${sid}`
  return sid
}

export function withRegion<T = void>(unit: NodeUnit, cb: () => T): T {
  const node = getGraph(unit)
  const meta = node.meta || {}

  if (!is.domain(unit)) {
    meta.regionStack = regionStack
  }

  regionStack = {
    id: node.id,
    parent: regionStack,
    value: unit,
    template: meta.template || readTemplate(),
    sidRoot: meta.sidRoot || (regionStack && regionStack.sidRoot),
    meta: meta,
  }
  try {
    return cb()
  } finally {
    reportDeclaration('region')
    regionStack = getParent(regionStack)
  }
}

export const withFactory = ({
  sid,
  name,
  loc,
  method,
  fn,
}: {
  sid: string
  name?: string
  loc?: any
  method?: string
  fn: () => any
}) => {
  const factoryRootNode = createNode({
    meta: {sidRoot: readSidRoot(sid), sid, name, loc, method, type: 'factory'},
    regional: true,
  })

  return withRegion(factoryRootNode, fn)
}
