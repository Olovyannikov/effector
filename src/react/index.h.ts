import type {Event, Store, EventCallable} from 'effector'
import type {ComponentType, ReactNode} from 'react'

export type StoreConsumer<State> = ComponentType<{
  children: (state: State) => ReactNode,
}>

export type StoreView<State, Props> = ComponentType<Props> & {
  mounted: Event<{props: Props, state: State}>,
  unmounted: Event<{props: Props, state: State}>,
}

export type Gate<Props = {}> = React.ComponentType<Props> & {
  open: EventCallable<Props>
  close: EventCallable<Props>
  status: Store<boolean>
  state: Store<Props>
  set: EventCallable<Props>
}
