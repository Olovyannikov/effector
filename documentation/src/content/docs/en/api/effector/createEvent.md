---
title: createEvent
description: createEvent is a method for creating an event
redirectFrom:
  - /api/effector/createEvent
  - /docs/api/effector/createEvent
---

# createEvent (#create-event)

```ts
import { createEvent } from "effector";

const event = createEvent();
```

Method for creating [events][eventApi].

## Formula (#formulae)

```ts
createEvent<E = void>(eventName?: string): EventCallable<E>
createEvent<E = void>(config: {
  name?: string
  sid?: string
  domain?: Domain
}): EventCallable<E>
```

- **Arguments**

  - `eventName`: Optional argument. Event name for debugging.
  - `config`: Optional argument. Configuration object.

    - `name`: Event name.
    - `sid`: Stable identifier for SSR.
    - `domain`: Domain for the event.

- **Return value**

Returns a new callable [event][eventTypes].

## Examples (#examples)

Updating state by calling an event:

```js
import { createStore, createEvent } from "effector";

const addNumber = createEvent();

const $counter = createStore(0);

$counter.on(addNumber, (state, number) => state + number);

$counter.watch((state) => {
  console.log("state", state);
});
// => 0

addNumber(10);
// => 10

addNumber(10);
// => 20

addNumber(10);
// => 30
```

[Run example](https://share.effector.dev/0OeoZMPc)

We created the `addNumber` event and the `$counter` store, then subscribed to store updates.<br/>
Notice the function call `addNumber(10)`. Every time you call `addNumber(10)`, you can check the console and see how the state changes.

Processing data with derived events:

```js
import { createEvent } from "effector";

const extractPartOfArray = createEvent();
const array = extractPartOfArray.map((arr) => arr.slice(2));

array.watch((part) => {
  console.log(part);
});
extractPartOfArray([1, 2, 3, 4, 5, 6]);
// => [3, 4, 5, 6]
```

[Run example](https://share.effector.dev/4lWsZr2k)

## Related API and Articles (#related-api-and-docs-to-create-event)

- **API**
  - [`Event API`][eventApi] - Event API, its methods, properties and description
  - [`createApi`][createApi] - Creating a set of events for a store
  - [`merge`][merge] - Method for combining an array of units into one new event
  - [`sample`][sample] - Connecting events with other units
- **Articles**
  - [How to work with events][eventGuide]
  - [How to think in effector and why events matter][mindset]
  - [Guide to typing events and other units][typescript]

[eventApi]: /en/api/effector/Event
[eventTypes]: /en/api/effector/Event#event-types
[merge]: /en/api/effector/merge
[eventGuide]: /en/essentials/events
[mindset]: /en/resources/mindset
[typescript]: /en/essentials/typescript
[sample]: /en/api/effector/sample
[createApi]: /en/api/effector/createApi
