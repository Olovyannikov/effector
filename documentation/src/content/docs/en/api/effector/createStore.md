---
title: createStore
description: createStore is a method for creating a store
redirectFrom:
  - /api/effector/createStore
  - /docs/api/effector/createStore
---

# createStore (#create-store)

```ts
import { createStore } from "effector";

const $store = createStore();
```

Method for creating [stores][storeApi].

## Formula (#createStore-formulae)

```ts
createStore(
  defaultState: State, // Initial store state
  config?: { // Configuration object with additional options
    skipVoid?: boolean; // Controls updates with undefined values
    name?: string; // Store name for debugging
    sid?: string // Stable identifier for SSR
    updateFilter?: (update: State, current: State) => boolean // Update filtering function
    serialize?: // Serialization configuration for SSR
    | 'ignore'
    | {
        write: (state: State) => SerializedState
        read: (json: SerializedState) => State
      }
    domain?: Domain; // Domain to which the store belongs
  },
): StoreWritable<State>
```

- **Arguments**

1. **`defaultState`**: Initial state
2. **`config`**: Optional configuration object

   - **`skipVoid`**: Optional argument. Determines whether the [store][storeApi] skips `undefined` values. Default is `true`. If you pass an `undefined` value to a store with `skipVoid: true`, you'll get [an error in the console][storeUndefinedError].<br/><br/>

   - **`name`**: Optional argument. Store name. [Babel-plugin][babel] can determine it from the store variable name if the name is not explicitly passed in the configuration.<br/><br/>
   - **`sid`**: Optional argument. Unique store identifier. [It's used to distinguish stores between different environments][storeSid]. When using [Babel-plugin][babel], it's set automatically.<br/><br/>
   - **`updateFilter`**:
     Optional argument. A [pure function][pureFn] that prevents store updates if it returns `false`. Should be used when the standard update prevention (if the value to be written to the store equals `undefined` or the current store value) is insufficient.

     <br/>

   - **`serialize`**: Optional argument responsible for store serialization.

     - `'ignore'`: excludes the store from serialization when calling [serialize][serialize].
     - Object with `write` and `read` methods for custom serialization. `write` is called when [serialize](/en/api/effector/serialize) is invoked and converts the store state to a JSON value – a primitive or simple object/array. `read` is called during [fork](/en/api/effector/fork) if the provided `values` are the result of calling [serialize][serialize].

- **Return value**

Returns a new [store][storeApi].

## Examples (#examples)

Basic store usage:

```js
import { createEvent, createStore } from "effector";

const addTodo = createEvent();
const clearTodos = createEvent();

const $todos = createStore([])
  .on(addTodo, (todos, newTodo) => [...todos, newTodo])
  .reset(clearTodos);

const $selectedTodos = $todos.map((todos) => {
  return todos.filter((todo) => !!todo.selected);
});

$todos.watch((todos) => {
  console.log("todos", todos);
});
```

[Run example](https://share.effector.dev/tquiUgdq)

Example with custom `serialize` configuration:

```ts
import { createEvent, createStore, serialize, fork, allSettled } from "effector";

const saveDate = createEvent();
const $date = createStore<null | Date>(null, {
  // Date objects are automatically converted to ISO date strings when calling JSON.stringify
  // but are not converted back to Date when calling JSON.parse – the result will be the same ISO date string
  // This will cause state mismatch when hydrating state on the client during server-side rendering
  //
  // Custom `serialize` configuration solves this problem
  serialize: {
    write: (dateOrNull) => (dateOrNull ? dateOrNull.toISOString() : dateOrNull),
    read: (isoStringOrNull) => (isoStringOrNull ? new Date(isoStringOrNull) : isoStringOrNull),
  },
}).on(saveDate, (_, p) => p);

const serverScope = fork();

await allSettled(saveDate, { scope: serverScope, params: new Date() });

const serverValues = serialize(serverScope);
// `serialize.write` for store `$date` was called

console.log(serverValues);
// => { nq1e2rb: "2022-11-05T15:38:53.108Z" }
// Date object from store saved as ISO date

const clientScope = fork({ values: serverValues });
// `serialize.read` for store `$date` was called

const currentDate = clientScope.getState($date);
console.log(currentDate);
// => Date 11/5/2022, 10:40:13 PM
// ISO date string converted back to Date object
```

[Run example](https://share.effector.dev/YFkUlqPv)

## Common Errors (#common-errors)

Below is a list of possible errors you may encounter when working with stores:

- [`store: undefined is used to skip updates. To allow undefined as a value provide explicit { skipVoid: false } option`][storeUndefinedError].
- [`serialize: One or more stores dont have sids, their values are omitted`][serializeError].
- [`unit call from pure function is not supported, use operators like sample instead`][unitCallError].

## Related API and Articles (#related-api-and-docs-to-create-store)

- **API**
  - [`Store API`][storeApi] - Store API, its methods, properties and description
  - [`createApi`][createApi] - Creating a set of events for a store
  - [`combine`][combine] - Creating a new store based on other stores
  - [`sample`][sample] - Connecting stores with other units
- **Articles**
  - [How to manage state][storeGuide]
  - [Guide to working with SSR][ssr]
  - [What is SID and why stores need them][storeSid]
  - [How to type stores and other units][typescript]

[storeApi]: /en/api/effector/Store
[storeUndefinedError]: /en/guides/troubleshooting#store-undefined
[storeSid]: /en/explanation/sids
[ssr]: /en/guides/server-side-rendering
[storeGuide]: /en/essentials/manage-states
[combine]: /en/api/effector/combine
[sample]: /en/api/effector/sample
[createApi]: /en/api/effector/createApi
[serialize]: /en/api/effector/serialize
[typescript]: /en/essentials/typescript
[babel]: /en/api/effector/babel-plugin
[pureFn]: /en/explanation/glossary/#purity
[unitCallError]: /en/guides/troubleshooting#unit-call-from-pure-not-supported
[serializeError]: /en/guides/troubleshooting/#store-without-sid
