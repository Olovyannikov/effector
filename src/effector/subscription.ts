import type {Subscription, Node} from './index.h'
import {clearNodeLight} from './clearNode'

export const createSubscription = (node: Node): Subscription =>
  addUnsubscribe(() => clearNodeLight(node))

export function addUnsubscribe(callback: () => void): Subscription {
  const subscription: Subscription = () => callback()
  subscription.unsubscribe = () => callback()

  return subscription
}
