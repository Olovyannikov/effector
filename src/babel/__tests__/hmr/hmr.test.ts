// @ts-expect-error no types
import tsPreset from '@babel/preset-typescript'

// @ts-expect-error no types
import babelPlugin from '../../babel-plugin'
import {compile, configSetup, readAndCompile} from '../utils'

describe('hmr module', () => {
  test('esm format', async () => {
    expect(
      await readAndCompile(__dirname, 'hmr.js', configSetup('es', 'js')),
    ).toMatchSnapshot()
  })

  test('cjs format', async () => {
    expect(
      await readAndCompile(__dirname, 'hmr.js', configSetup('cjs', 'js')),
    ).toMatchSnapshot()
  })
})

test('dont throw on already imported methods', async () => {
  const code = `
  import {createNode, withRegion} from 'effector'
  `
  expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(
    `""`,
  )
})

test('reference with factories', async () => {
  const code = `
  import type {Store} from 'effector'
  import {debug} from 'patronum'

  debug()
  `
  expect(
    await compile(code, {
      plugins: [[babelPlugin, {addNames: false, addLoc: false}]],
      presets: [[tsPreset, {isTSX: true, allExtensions: true}]],
      configFile: false,
      babelrc: false,
    }),
  ).toMatchInlineSnapshot(`
    "import {withFactory as _withFactory} from 'effector'
    import {debug} from 'patronum'

    _withFactory({
      sid: 'ujj0q3',
      fn: () => debug(),
    })
    "
  `)
})

describe('put hmr only to modules with effector', () => {
  test('no effector no hmr', async () => {
    const code = `import {useMemo} from 'react'`
    expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(
      `""`,
    )
  })

  test('has factories has hmr', async () => {
    const code = `
    import {debug} from 'patronum'

    debug()
    `
    expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(`
      "import {
        createNode as _createNode,
        clearNode as _clearNode,
        withRegion as _withRegion,
        withFactory as _withFactory,
      } from 'effector'

      const _internalHMRRegion = _createNode({
        regional: true,
      })

      import {debug} from 'patronum'

      _withRegion(_internalHMRRegion, () =>
        _withFactory({
          sid: 'uiz86g',
          fn: () => debug(),
        }),
      )

      if (import.meta.hot || import.meta.webpackHot) {
        ;(import.meta.hot || import.meta.webpackHot).dispose(() =>
          _clearNode(_internalHMRRegion),
        )
      } else {
        console.warn('[effector hmr] HMR is not available in current environment.')
      }
      "
    `)
  })

  test('has stores has hmr', async () => {
    const code = `
    import {createStore} from 'effector'
    const $foo = createStore(0)
    `
    expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(`
      "import {
        createNode as _createNode,
        clearNode as _clearNode,
        withRegion as _withRegion,
      } from 'effector'

      const _internalHMRRegion = _createNode({
        regional: true,
      })

      import {createStore} from 'effector'

      const $foo = _withRegion(_internalHMRRegion, () =>
        createStore(0, {
          sid: 'fjm8r',
        }),
      )

      if (import.meta.hot || import.meta.webpackHot) {
        ;(import.meta.hot || import.meta.webpackHot).dispose(() =>
          _clearNode(_internalHMRRegion),
        )
      } else {
        console.warn('[effector hmr] HMR is not available in current environment.')
      }
      "
    `)
  })

  test('unused imports', async () => {
    const code = `
    import {createStore} from 'effector'
    import {useMemo} from 'react'
    `
    expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(
      `""`,
    )
  })
})

describe('imports correctness', () => {
  test('type imports', async () => {
    const code = `
    import type {Store} from 'effector'
    import {debug} from 'patronum'
    `
    expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(
      `""`,
    )
  })
  test('imports in the end of file', async () => {
    const code = `
    import {debug} from 'patronum'
    debug()
    import {createStore} from 'effector'
    const $foo = createStore(0)
    `
    expect(await compile(code, configSetup('es', 'ts'))).toMatchInlineSnapshot(`
      "import {
        createNode as _createNode,
        clearNode as _clearNode,
        withRegion as _withRegion,
        withFactory as _withFactory,
      } from 'effector'

      const _internalHMRRegion = _createNode({
        regional: true,
      })

      import {debug} from 'patronum'

      _withRegion(_internalHMRRegion, () =>
        _withFactory({
          sid: 'uiffl3',
          fn: () => debug(),
        }),
      )

      import {createStore} from 'effector'

      const $foo = _withRegion(_internalHMRRegion, () =>
        createStore(0, {
          sid: '1dmv15',
        }),
      )

      if (import.meta.hot || import.meta.webpackHot) {
        ;(import.meta.hot || import.meta.webpackHot).dispose(() =>
          _clearNode(_internalHMRRegion),
        )
      } else {
        console.warn('[effector hmr] HMR is not available in current environment.')
      }
      "
    `)
  })
})
