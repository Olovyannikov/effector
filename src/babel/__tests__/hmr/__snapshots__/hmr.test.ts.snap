// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hmr module cjs format 1`] = `
"'use strict'

Object.defineProperty(exports, '__esModule', {
  value: true,
})
exports.default = exports.exportedLet = exports.exportedConst = void 0

var _effector = require('effector')

var _factories = require('@withease/factories')

const _internalHMRRegion = (0, _effector.createNode)({
  regional: true,
})

function incorrect() {
  const $si = (0, _effector.createStore)(0, {
    sid: 's50ivo',
  })
  const ei = (0, _effector.createEvent)({
    sid: 'vcjkqm',
  })
  const iFx = (0, _effector.createEffect)(() => {}, {
    sid: '4szayb',
  })
  return {
    $si,
    ei,
    iFx,
  }
}

const incorrectArrow = () =>
  (0, _effector.createStore)(0, {
    sid: 'kf1ayc',
  })

const correct = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.withFactory)({
    sid: '-ikkxki',
    fn: () =>
      (0, _factories.createFactory)(() => {
        const $si = (0, _effector.createStore)(0, {
          sid: '-ybsckh',
        })
        const ei = (0, _effector.createEvent)({
          sid: '-k2ug0b',
        })
        const iFx = (0, _effector.createEffect)(() => {}, {
          sid: '-4raw84',
        })
        return {
          $si,
          ei,
          iFx,
        }
      }),
  }),
)
const fx = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.createEffect)(() => {}, {
    sid: 'jvh2by',
  }),
)
const attachedFx = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.attach)({
    and: {
      effect: fx,
    },
    or: {
      sid: 'rpnd5t',
    },
  }),
)
const $count = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.createStore)(0, {
    sid: 's9hq26',
  }),
)
const increment = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.createEvent)({
    sid: 'mq1bc6',
  }),
)
const decrement = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.createEvent)({
    sid: '-ue8y8v',
  }),
)
const exportedConst = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.createStore)('', {
    sid: 'kijuqq',
  }),
)
exports.exportedConst = exportedConst
let exportedLet = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.createStore)('', {
    sid: 'ajlqjc',
  }),
)
exports.exportedLet = exportedLet
const model = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.withFactory)({
    sid: 'ueg7w5',
    fn: () => (0, _factories.invoke)(correct),
  }),
)
const incorrectModel = incorrect()
const incorrectArrowModel = incorrectArrow()
;(0, _effector.withRegion)(_internalHMRRegion, () =>
  $count.on(increment, count => count + 1),
)
;(0, _effector.withRegion)(_internalHMRRegion, () => $count.watch(console.log))
const $store = (0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.sample)({
    and: [
      {
        clock: $count,
      },
    ],
    or: {
      sid: 'x4yuah',
    },
  }),
)

if (true) {
  ;(0, _effector.withRegion)(_internalHMRRegion, () =>
    (0, _effector.sample)({
      and: [
        {
          clock: increment,
          source: $count,
          fn: count => count + 1,
          target: $count,
        },
      ],
      or: {
        sid: 'n1dbhi',
      },
    }),
  )
}

;(0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.sample)({
    and: [
      {
        clock: decrement,
        source: $count,
        fn: count => count - 1,
        target: $count,
      },
    ],
    or: {
      sid: 'nhbcn9',
    },
  }),
)
;(0, _effector.withRegion)(_internalHMRRegion, () =>
  (0, _effector.sample)({
    and: [
      {
        clock: decrement,
        target: increment.prepend(() => -1),
      },
    ],
    or: {
      sid: 'nl5wss',
    },
  }),
)

class Class {
  event = (0, _effector.createEvent)({
    sid: 'xp6hes',
  })
}

var _default = (0, _effector.createEvent)({
  sid: 'z4bcpo',
})

exports.default = _default

if (module.hot) {
  module.hot.dispose(() => (0, _effector.clearNode)(_internalHMRRegion))
} else {
  console.warn('[effector hmr] HMR is not available in current environment.')
}
"
`;

exports[`hmr module esm format 1`] = `
"import {
  createNode as _createNode,
  clearNode as _clearNode,
  withRegion as _withRegion,
  withFactory as _withFactory,
} from 'effector'

const _internalHMRRegion = _createNode({
  regional: true,
})

import {createEvent, createStore, sample, createEffect, attach} from 'effector'
import {createFactory, invoke} from '@withease/factories'

function incorrect() {
  const $si = createStore(0, {
    sid: 's50ivo',
  })
  const ei = createEvent({
    sid: 'vcjkqm',
  })
  const iFx = createEffect(() => {}, {
    sid: '4szayb',
  })
  return {
    $si,
    ei,
    iFx,
  }
}

const incorrectArrow = () =>
  createStore(0, {
    sid: 'kf1ayc',
  })

const correct = _withRegion(_internalHMRRegion, () =>
  _withFactory({
    sid: '-ikkxki',
    fn: () =>
      createFactory(() => {
        const $si = createStore(0, {
          sid: '-ybsckh',
        })
        const ei = createEvent({
          sid: '-k2ug0b',
        })
        const iFx = createEffect(() => {}, {
          sid: '-4raw84',
        })
        return {
          $si,
          ei,
          iFx,
        }
      }),
  }),
)

const fx = _withRegion(_internalHMRRegion, () =>
  createEffect(() => {}, {
    sid: 'jvh2by',
  }),
)

const attachedFx = _withRegion(_internalHMRRegion, () =>
  attach({
    and: {
      effect: fx,
    },
    or: {
      sid: 'rpnd5t',
    },
  }),
)

const $count = _withRegion(_internalHMRRegion, () =>
  createStore(0, {
    sid: 's9hq26',
  }),
)

const increment = _withRegion(_internalHMRRegion, () =>
  createEvent({
    sid: 'mq1bc6',
  }),
)

const decrement = _withRegion(_internalHMRRegion, () =>
  createEvent({
    sid: '-ue8y8v',
  }),
)

export const exportedConst = _withRegion(_internalHMRRegion, () =>
  createStore('', {
    sid: 'kijuqq',
  }),
)
export let exportedLet = _withRegion(_internalHMRRegion, () =>
  createStore('', {
    sid: 'ajlqjc',
  }),
)

const model = _withRegion(_internalHMRRegion, () =>
  _withFactory({
    sid: 'ueg7w5',
    fn: () => invoke(correct),
  }),
)

const incorrectModel = incorrect()
const incorrectArrowModel = incorrectArrow()

_withRegion(_internalHMRRegion, () => $count.on(increment, count => count + 1))

_withRegion(_internalHMRRegion, () => $count.watch(console.log))

const $store = _withRegion(_internalHMRRegion, () =>
  sample({
    and: [
      {
        clock: $count,
      },
    ],
    or: {
      sid: 'x4yuah',
    },
  }),
)

if (true) {
  _withRegion(_internalHMRRegion, () =>
    sample({
      and: [
        {
          clock: increment,
          source: $count,
          fn: count => count + 1,
          target: $count,
        },
      ],
      or: {
        sid: 'n1dbhi',
      },
    }),
  )
}

_withRegion(_internalHMRRegion, () =>
  sample({
    and: [
      {
        clock: decrement,
        source: $count,
        fn: count => count - 1,
        target: $count,
      },
    ],
    or: {
      sid: 'nhbcn9',
    },
  }),
)

_withRegion(_internalHMRRegion, () =>
  sample({
    and: [
      {
        clock: decrement,
        target: increment.prepend(() => -1),
      },
    ],
    or: {
      sid: 'nl5wss',
    },
  }),
)

class Class {
  event = createEvent({
    sid: 'xp6hes',
  })
}

export default createEvent({
  sid: 'z4bcpo',
})

if (import.meta.hot || import.meta.webpackHot) {
  ;(import.meta.hot || import.meta.webpackHot).dispose(() =>
    _clearNode(_internalHMRRegion),
  )
} else {
  console.warn('[effector hmr] HMR is not available in current environment.')
}
"
`;
